File: binary-tree-postorder-traversal/Solution145.java
// https://leetcode.com/problems/binary-tree-postorder-traversal/

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class Solution145 {

    // T: O(n), S: O(n)
    public List<Integer> postorderTraversal(TreeNode root) {
        var result = new ArrayList<Integer>();
        helper(root, result);
        return result;
    }

    private void helper(TreeNode node, List<Integer> result) {
        if (node != null) {
            helper(node.left, result);
            helper(node.right, result);
            result.add(node.val);
        }
    }

    // T: O(n), S: O(n)
    public List<Integer> postorderTraversalIterative(TreeNode root) {
        var result = new ArrayList<Integer>();
        if (root == null) {
            return result;
        }
        var s1 = new Stack<TreeNode>();
        var s2 = new Stack<TreeNode>();
        s1.push(root);
        while (!s1.isEmpty()) {
            var curr = s1.pop();
            s2.push(curr);

            if (curr.left != null) {
                s1.push(curr.left);
            }
            if (curr.right != null) {
                s1.push(curr.right);
            }
        }
        while (!s2.isEmpty()) {
            result.add(s2.pop().val);
        }
        return result;
    }

    public static void main(String[] args) {
        var sol = new Solution145();
        // [1,null,2,3]
        var root1 = new TreeNode(1);
        root1.right = new TreeNode(2);
        root1.right.left = new TreeNode(3);
        System.out.println(sol.postorderTraversal(root1)); // [3,2,1]
        System.out.println(sol.postorderTraversalIterative(root1)); // [3,2,1]
    }

}

File: binary-tree-postorder-traversal/solution145.py
# https://leetcode.com/problems/binary-tree-postorder-traversal/

from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution145:

    # T: O(n), S: O(n)
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []

        def helper(node):
            if node:
                helper(node.left)
                helper(node.right)
                result.append(node.val)
        helper(root)
        return result

    # T: O(n), S: O(n)
    def postorderTraversalIterative(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        if not root:
            return result
        s1 = [root]
        s2 = []
        while s1:
            curr = s1.pop()
            s2.append(curr)
            if curr.left:
                s1.append(curr.left)
            if curr.right:
                s1.append(curr.right)
        while s2:
            result.append(s2.pop().val)
        return result


def main() -> None:
    sol = Solution145()
    root1 = TreeNode(1)
    root1.right = TreeNode(2)
    root1.right.left = TreeNode(3)
    print(sol.postorderTraversal(root1))         # [3, 2, 1]
    print(sol.postorderTraversalIterative(root1))  # [3, 2, 1]


if __name__ == "__main__":
    main()
